pkg/agent/multicast/vif_allocator.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	multicastsyscall "antrea.io/antrea/third_party/unix"
+
+	"k8s.io/klog/v2"
+)
+
+// VIF stands for virtual interface for multicast in kernel multicast implementation
+// and is used for adding/deleting static multicast entries, please check
+// https://github.com/torvalds/linux/blob/master/include/uapi/linux/mroute.h#L80
+// VIFAllocator is used for saving vif <-> interface mapping.
+type VIFAllocator struct {
+	mutex  sync.RWMutex
+	vifMap map[string]uint16
+	sockFD int
+}
+
+func newVIFAllocator(sockFD int) *VIFAllocator {
+	return &VIFAllocator{
+		vifMap: make(map[string]uint16),
+		sockFD: sockFD,
+	}
+}
+
+func (v *VIFAllocator) getVIF(name string) (uint16, error) {
+	v.mutex.RLock()
+	defer v.mutex.RUnlock()
+	vif, ok := v.vifMap[name]
+	if !ok {
+		return 0, fmt.Errorf("VIF for interface %s not found", name)
+	}
+	return vif, nil
+}
+
+func (v *VIFAllocator) allocateVIFs(names []string) ([]uint16, error) {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return []uint16{}, err
+	}
+	vifs := make([]uint16, 0)
+	for _, iface := range ifaces {
+		for _, name := range names {
```

Comment:
I think the outer looper should be names as we are trying to allocate VIFs for names, ifaces is just used to get index of specific name.

Reply:
Actually we are trying to allocate VIFs for interface index , not names. Check this kernel file. https://elixir.bootlin.com/linux/v5.15/source/include/uapi/linux/mroute.h#L61 . We call `secksockopt` with `vifctl`, and interface index is used in `vifc_lcl_ifindex` field.    An alternative solution would be allocating VIFs for interface address. 

Actually if we don't care about logging, we do not need to store interface names in multicast route code because interface names are not used by any multicast syscalls

Reply:
updated. Looks better in this way, thanks.